tests/integration/bluez_controller.py

```diff
@@ -1,83 +1,120 @@
+import sys
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    if sys.platform != "linux":
+        assert False, "This is only available on Linux"
+
+
 import asyncio
 import contextlib
 import logging
-from typing import AsyncGenerator, cast
+from typing import AsyncGenerator
 
-from bumble import hci
 from bumble.controller import Controller
 from bumble.link import LocalLink
 from bumble.transport import open_transport
+from bumble.transport.common import Transport
+from dbus_fast import BusType, Message, MessageType, Variant
+from dbus_fast.aio.message_bus import MessageBus
 
-EMPTY_ADDRESS = hci.Address("00:00:00:00:00:00")
-
-
-def log_output(output: bytes, prefix: str = "") -> None:
-    """Log subprocess output line by line."""
-    for line in output.decode(errors="ignore").splitlines():
-        logging.debug(f"{prefix}{line}")
-
-
-async def run_bluetoothctl(commands: list[str]) -> tuple[int, str, str]:
-    """Run bluetoothctl with the given commands."""
-    bt = await asyncio.create_subprocess_exec(
-        "bluetoothctl",
-        stdin=asyncio.subprocess.PIPE,
-        stdout=asyncio.subprocess.PIPE,
-        stderr=asyncio.subprocess.PIPE,
-    )
+from bleak.backends.bluezdbus import defs
+from bleak.backends.bluezdbus.signals import MatchRules, add_match
+from bleak.backends.bluezdbus.utils import assert_reply, get_dbus_authenticator
 
-    commands_str = "\n".join(commands) + "\n"
-    stdout, stderr = await bt.communicate(input=commands_str.encode())
+if sys.version_info < (3, 11):
+    from async_timeout import timeout as async_timeout
+else:
+    from asyncio import timeout as async_timeout
 
-    return (
-        bt.returncode or 0,
-        stdout.decode(errors="ignore"),
-        stderr.decode(errors="ignore"),
-    )
-
-
-async def controller_available(address: hci.Address) -> bool:
-    """Check if a Bluetooth controller is available via bluetoothctl show."""
-    commands = [f"show {address}", "quit"]
-    _, stdout, _ = await run_bluetoothctl(commands)
-
-    if "Manufacturer: 0xffff" in stdout:
-        return True
-    return False
 
+@contextlib.asynccontextmanager
+async def open_message_bus() -> AsyncGenerator[MessageBus, None]:
+    bus = MessageBus(bus_type=BusType.SYSTEM, auth=get_dbus_authenticator())
+    await bus.connect()
+    try:
+        yield bus
+    finally:
+        if bus.connected:
+            bus.disconnect()
```

Comment:
Actually `disconnect()` should be called unconditionally, and even if `connect()` fails.

It probably doesn't make much difference for tests, but people might copy this to "real" code, so best to do it right.